pandevice/predefined.py

```diff
@@ -0,0 +1,269 @@
+
+        elm = root.find("result")
+        return elm
+
+    def _parse_application_xml(self, xml):
+        """parse the xml into actual objects and store them in the dicts"""
+        
+        for elm in xml:
+            if elm.find("functions") is not None:
+                # this is an ApplicationContainerObject
+                obj = objects.ApplicationContainer()
+                obj.refresh(xml=elm)
+                self.application_container_objects[obj.name] = obj
+            else:
+                # this is an ApplicaitonObject
+                obj = objects.ApplicationObject()
+                obj.refresh(xml=elm)
+                self.application_objects[obj.name] = obj
+    
+    def _parse_service_xml(self, xml):
+        """parse the xml into actual objects and store them in the dicts"""
+        
+        for elm in xml:
+            obj = objects.ServiceObject()
+            obj.refresh(xml=elm)
+            self.service_objects[obj.name] = obj
+
+    def refresh_application(self, name):
+        """Refesh a Single Predefined Applciation
+
+        This method refreshes single predefined applciation or application container 
+        (predefined only object). 
+
+        Args:
+            name (str): Name of the application to refresh
+        
+        """
+
+        xpath = self.SINGLE_APPLICATION_XPATH % name
+        xml = self._get_xml(xpath)
+        self._parse_application_xml(xml)
+
+    def refresh_service(self, name):
+        """Refesh a Single Predefined Service
+
+        This method refreshes single predefined service (predefined only object). 
+
+        Args:
+            name (str): Name of the service to refresh
+        
+        """
+
+        xpath = self.SINGLE_SERVIE_XPATH % name
+        xml = self._get_xml(xpath)
+        self._parse_service_xml(xml)
+
+    def refreshall_applications(self):
+        """Refesh all Predefined Applications
+
+        This method refreshes all predefined applications and application containers.
+        
+        CAUTION: This method requires a lot of overhead on the device api to respond.
+        It will like some time (1-2+ minutes to respond usually).
+        
+        """
+
+        xpath = self.ALL_APPLICATION_XPATH + "/entry"
+        xml = self._get_xml(xpath)
+        self._parse_application_xml(xml)
+
+    def refreshall_services(self):
+        """Refesh all Predefined Services
+
+        This method refreshes all predefined services.
+        
+        """
+
+        xpath = self.ALL_SERVICE_XPATH + "/entry"
+        xml = self._get_xml(xpath)
+        self._parse_service_xml(xml)
+
+    def refreshall(self):
+        """Refesh all Predefined Objects
+
+        This method refreshes all predefined objects. This includes applications,
+        applciation containerd, and services.
+
+        CAUTION: This method requires a lot of overhead on the device api to respond.
+        It will like some time (1-2+ minutes to respond usually).
```

Comment:
Interesting, I have been testing on a PA-3020 and a virtual panorama and both have been around 75 seconds. That of course is likely my machine processing the response though. I will change the wording to match what you said in that "your mileage may vary" depending on PAN platform and circumstances.